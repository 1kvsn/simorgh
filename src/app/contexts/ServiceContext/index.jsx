import React from 'react';
import { node, string } from 'prop-types';
import services from '#lib/config/services/loadableConfig';
import variantPropType from '../../models/propTypes/variants';
import { getVariant } from '#lib/utilities/variantHandler';
import getLangOverride from '#lib/utilities/langHandler';

export const ServiceContext = React.createContext({});

export const ServiceContextProvider = ({
  children,
  service,
  variant,
  pageLang,
}) => {
  const LoadableContextProvider = services[service];

  if (!LoadableContextProvider) {
    return null;
  }

  return (
    <LoadableContextProvider
      Context={ServiceContext}
      dataKey={getLangOverride(pageLang) || getVariant({ service, variant })}
    >
      {children}
    </LoadableContextProvider>
  );
};

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
  variant: variantPropType,
  pageLang: string,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
  variant: 'default',
  pageLang: null,
};
