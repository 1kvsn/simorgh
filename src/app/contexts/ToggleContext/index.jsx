import React, { createContext, useReducer } from 'react';
import { node, object } from 'prop-types';
import { toggleReducer } from './reducer';

const ToggleContext = createContext({});

const ToggleContextProvider = ({ children, toggles }) => {
  const [toggleState, toggleDispatch] = useReducer(toggleReducer, toggles);

  return (
    <ToggleContext.Provider value={{ toggleState, toggleDispatch }}>
      {children}
    </ToggleContext.Provider>
  );
};

const ToggleContextConsumer = ToggleContext.Consumer;

ToggleContextProvider.propTypes = {
  children: node.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  toggles: object.isRequired,
};

export { ToggleContext, ToggleContextProvider, ToggleContextConsumer };
