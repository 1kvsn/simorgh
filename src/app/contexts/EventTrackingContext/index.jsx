import React, { createContext, useContext } from 'react';
import { node } from 'prop-types';

import { pageDataPropType } from '#models/propTypes/data';
import { ServiceContext } from '#contexts/ServiceContext';
import { RequestContext } from '#contexts/RequestContext';
import { buildATIClickParams } from '#containers/ATIAnalytics/params';
import useToggle from '#hooks/useToggle';
import {
  ARTICLE_PAGE,
  FRONT_PAGE,
  MEDIA_PAGE,
  MOST_READ_PAGE,
  MOST_WATCHED_PAGE,
  INDEX_PAGE,
  FEATURE_INDEX_PAGE,
  MEDIA_ASSET_PAGE,
  STORY_PAGE,
  PHOTO_GALLERY_PAGE,
  CORRESPONDENT_STORY_PAGE,
} from '#app/routes/utils/pageTypes';

export const EventTrackingContext = createContext({});

export const EventTrackingContextProvider = ({ children, pageData }) => {
  let pageIdentifier;
  let platform;
  let statsDestination;

  const requestContext = useContext(RequestContext);
  const serviceContext = useContext(ServiceContext);
  const { enabled: eventTrackingIsEnabled } = useToggle('eventTracking');

  if (pageData && eventTrackingIsEnabled) {
    try {
      ({ pageIdentifier, platform, statsDestination } = buildATIClickParams(
        pageData,
        requestContext,
        serviceContext,
      ));
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(
        `ATI Event Tracking Error: Could not parse tracking values from page data:\n${error.message}`,
      );
    }
  }

  const { pageType } = requestContext;
  const campaignName = {
    [ARTICLE_PAGE]: 'article',
    [FRONT_PAGE]: 'index-home',
    [MEDIA_PAGE]: 'player-episode-tv',
    [MOST_READ_PAGE]: 'list-datadriven-read',
    [MOST_WATCHED_PAGE]: 'list-datadriven-watched',
    [INDEX_PAGE]: 'index-section',
    [FEATURE_INDEX_PAGE]: 'index-section-fix',
    [MEDIA_ASSET_PAGE]: 'article-media-asset',
    [STORY_PAGE]: 'article-sty',
    [PHOTO_GALLERY_PAGE]: 'article-photo-gallery',
    [CORRESPONDENT_STORY_PAGE]: 'article-csp',
  }[pageType];

  if (!campaignName && pageData) {
    // eslint-disable-next-line no-console
    console.error(
      `ATI Event Tracking Error: Could not get the page type's campaign name`,
    );
  }

  const hasRequiredProps = [
    campaignName,
    pageIdentifier,
    platform,
    statsDestination,
  ].every(Boolean);
  const trackingData = hasRequiredProps
    ? {
        campaignName,
        pageIdentifier,
        platform,
        statsDestination,
      }
    : {};

  return (
    <EventTrackingContext.Provider value={trackingData}>
      {children}
    </EventTrackingContext.Provider>
  );
};

EventTrackingContextProvider.propTypes = {
  children: node.isRequired,
  pageData: pageDataPropType,
};

EventTrackingContextProvider.defaultProps = {
  pageData: null,
};
