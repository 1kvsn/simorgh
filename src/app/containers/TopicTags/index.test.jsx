import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import { ServiceContextProvider } from '#contexts/ServiceContext';
import { ToggleContextProvider } from '#contexts/ToggleContext';
import { EventTrackingContextProvider } from '#contexts/EventTrackingContext';
import Topics from '#containers/TopicTags';

const withContexts = ({ children }) => (
  <ServiceContextProvider service="pidgin">
    <ToggleContextProvider
      toggles={{
        eventTracking: {
          enabled: true,
        },
      }}
    >
      <EventTrackingContextProvider>{children}</EventTrackingContextProvider>
    </ToggleContextProvider>
  </ServiceContextProvider>
);

describe('TopicTags', () => {
  // Different variations match snapshots
  // Should call event tracking hooks

  shouldMatchSnapshot(
    'should render correctly with no tags',
    withContexts(<Topics />),
  );

  shouldMatchSnapshot(
    'should render correctly with a single tag',
    withContexts(<Topics topics={[{ topicName: 'topic', topicId: '123' }]} />),
  );

  shouldMatchSnapshot(
    'should render correctly with multiple tags',
    withContexts(
      <Topics
        topics={[
          { topicName: 'topic1', topicId: '1' },
          { topicName: 'topic2', topicId: '2' },
          { topicName: 'topic3', topicId: '3' },
        ]}
      />,
    ),
  );
});
