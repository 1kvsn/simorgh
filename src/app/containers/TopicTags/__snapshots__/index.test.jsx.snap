// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopicTags should render correctly with a single tag 1`] = `<div />`;

exports[`TopicTags should render correctly with multiple tags 1`] = `<div />`;

exports[`TopicTags should render correctly with no tags 1`] = `<div />`;
